course3/homework-list/list.h

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <memory>
+
+
+namespace Karun
+{
+
+template<typename T>
+class List
+{
+public:
+	class Node
+	{
+	friend class List;
+	
+	public:
+		Node(const T& data, std::unique_ptr<Node>& next)
+		: 	mData(data)
+		{
+			mNext = std::move(next);
+		}
+		
+	private:
+		T mData;
+		std::unique_ptr<Node> mNext;
+	};
+	
+	class iterator
+	{
+	friend class List;
+		
+	public:
+		iterator(Node* node) : mNode(node) {}
+		
+		bool operator==(const iterator& it) { return mNode == it.mNode; }
+		bool operator!=(const iterator& it) { return !operator==(it); }
+		iterator& operator++() { mNode = mNode->mNext.get(); return *this; }
+		T& operator*() { return mNode->mData; }
+		
+	private:
+		Node* mNode = nullptr;
+	};
+	
+	class const_iterator
+	{
+	public:
+		const_iterator(const iterator& it) : mNode(it.mNode) {}
+		const_iterator(Node* node) : mNode(node) {}
+		
+		bool operator==(const const_iterator& it) { return mNode == it.mNode; }
+		bool operator!=(const const_iterator& it) { return !operator==(it); }
+		const_iterator& operator++() { mNode = mNode->mNext.get(); return *this; }
+		const T& operator*() { return mNode->mData; }
+		
+	private:
+		Node* mNode = nullptr;
+	};
+	
+	List() = default;
+	List(const std::initializer_list<T>& list) { for (const T& t : list) push_front(t); }
```

Comment:
The initializer list works well in general, but unfortunately seems to be incompatible with move semantics. As a result, I cannot use this implementation to create a list of unique_ptrs.

More info on this here (I didn't try any of the fancy workarounds):
http://stackoverflow.com/a/8468817